### PR TITLE
KEYCLOAK-10112: Issues in loading offline session in a cluster environment during startup

### PR DESCRIPTION
**Pull request by rmartinc**
Fix for KEYCLOAK-10112. The PR adds the ability to lock the database using different namespaces (or lock types). It also contains some improvements for the offline session loading (avoid re-trying successful segments and using the `cardinality` in the `BitSet`).

I have just added some tests for the new lock namespaces. The rest of the tests should be there before.

### CODE REVIEW

**Comment by mposolda** [comment]
@rmartinc Could you please rebase the PR against latest master (I want to run DB matrix with your PR and there were some DB errors fixed in the meantime in the master)

**Comment by rmartinc** [comment]
@mposolda Rebased now!

**Comment by rmartinc** [comment]
@mposolda How is this going? Just let me know if you need something else in the PR.

**Comment by mposolda** [comment]
@rmartinc I am sorry, was on summit last week and still syncing a bit. Will take a look ASAP

File: server-spi-private/src/main/java/org/keycloak/models/dblock/DBLockProvider.java
Comment (mposolda) [inline server-spi-private/src/main/java/org/keycloak/models/dblock/DBLockProvider.java]: It seems a bit unclear that waitForLock has namespace argument, but for the other methods like "releaseLock" or "hasLock" it is not very clear how they work now.

It seems that with the liquibase implementation, you can't have "nested" lock (EG. obtain lock for namespace GENERAL and then have same node to obtain the lock also for OFFLINE_SESSIONS namespace when GENERAL is still held). Is it possible to make it clear in the javadoc? Also is it possible to rename the "hasLock" to "getCurrentLock", which will return null or namespace of the currently retrieved lock? For "releaseLock", I guess it will be good to not change signature, but just add the information that it is used to release the lock, which is currently held. WDYT?

**Review by mposolda — COMMENTED** [review]
@rmartinc Thanks for the PR and I am sorry for longer response. It's very good work and I like the changes in the offline sessions preloading part.

But I have some doubts about some liquibase lock changes and left the comments. Besides that, I've tried to run the DB matrix with all our supported databases and will let you know how it looks. But anyway, my vote is to remove the deletion of rows and make the javadocs more clear (and possibly change some method signatures to have the SPI more clear how it works). Thanks!

File: server-spi-private/src/main/java/org/keycloak/models/dblock/DBLockProvider.java
Comment (rmartinc) [reply server-spi-private/src/main/java/org/keycloak/models/dblock/DBLockProvider.java]: @mposolda I didn't think about nested locks... But I think that the "hasLock" method has the meaning if this node has the lock (previously the only one that was possible). So it should be "hasLock(Namespace)" for example is used [here](https://github.com/keycloak/keycloak/blob/master/model/jpa/src/main/java/org/keycloak/connections/jpa/DefaultJpaConnectionProviderFactory.java#L330) without specifying any "waitForLock" before. Maybe it's better if we add the Namespace to the three methods (waitForLock, releaseLock and hasLock). WDYT?

Comment (rmartinc) [reply server-spi-private/src/main/java/org/keycloak/models/dblock/DBLockProvider.java]: No, the "releaseLock" should be done for the previous obtained lock. So the summary is: "waitForLock" and "hasLock" receive the namespace (the first method performs the select for update to the ID, the second just obtains if there is a lock from a concurrent set in the provider); the "releaseLock" rollbacks the previous select for update. When you get a "DBLockProvider" you can only waitForLock/releaseLock one namespace, if you need nested you need two providers (two connections).

I'm going to prepare the changes, let's see if I can commit something the next week.

**Comment by rmartinc** [comment]
@mposolda The two changes were added to the PR.

The First one was easy, now the IDs that are missing are just added (no deletes involved).

But for the second modification I spent long hours last Friday. The main problem is that the previous interpretation of the "hasLock" method was global, for example in [DefaultJpaConnectionProviderFactory|https://github.com/keycloak/keycloak/blob/master/model/jpa/src/main/java/org/keycloak/connections/jpa/DefaultJpaConnectionProviderFactory.java#L329]. So any DBLockProvider instance knew if there was a lock in the node (an atomic was used in the LiquibaseDBLockProviderFactory to know that). But I think this idea can be misleading (for example one thread obtains the lock and another thread checks the "hasLock" and thinks it has the lock, but later the lock is released in the middle of its task). Until your comment I thought that the "hasLock" was just local to the current provider. Finally I decided to follow this local interpretation, as you asked me to do in the PR. The renamed "getCurrentLock" only returns the namespace that was locked previously (locally in this provider, not global anymore) or null. This way two different threads in the same JVM can lock each other (as previously, but the "hasLock" had that different global meaning which was very strange IMO), and nested locks are also available (but using different providers/sessions).

As a consequence I needed to add a third namespace, now there are three: DATABASE(1), for DefaultJpaConnectionProviderFactory create/migrate; KEYCLOAK_BOOT(1000), used in KeycloakApplication; and OFFLINE_SESSIONS(1001), for loading offline sessions. My feeling is that the previous global "hasLock" interpretation was needed by the usage in DefaultJpaConnectionProviderFactory, now with a new namespace no special meaning is needed.

I also rebased to master again.

**Comment by mposolda** [comment]
@rmartinc Thanks! I think the changes you made makes sense.

I am trying to run DB matrix tests and test your PR with various databases. IMO PR is fine to merge once it's ok with all dbs.

**Comment by mposolda** [comment]
@rmartinc Summary of the DB run: MySQL and MariaDB passed. MSSQL test is failing anyway ATM due the other issue ( KEYCLOAK-10075 ), so that's OK with respect to your PR. PostgreSQL is also failing due the KEYCLOAK-10075, but with your branch, I saw some more test failures. This is probably unrelated to changes in your PR, but I need to doublecheck. For the MariaDB Galera, there are test failures in DBLockTest, however the test was failing before due the KEYCLOAK-10262 . I would like to take a look at this one first.

In summary, I would like to fix KEYCLOAK-10262 and then doublecheck that MariaDB Galera is passing with your PR added to that. Also will be good to fix KEYCLOAK-10075 first to doublecheck your PR doesn't add more regressions to PostgreSQL. Temporarily switching this PR to "On hold" state until KEYCLOAK-10262 and KEYCLOAK-10075 are passing.

**Comment by rmartinc** [comment]
Perfect @mposolda, I did my manual tests with postgresql and the DBLockTest works for sure in the database, so I think that postgresql is OK. For Galera I have no idea, I haven't tested anything with it. Let's see.

**Comment by mposolda** [comment]
@rmartinc Hi, for the fix of https://issues.jboss.org/browse/KEYCLOAK-10262 to have mariaDB Galera tests passing, I needed to do small changes in the DBLock implementation. See here for the details:  https://issues.jboss.org/browse/KEYCLOAK-10262?focusedCommentId=13741310&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-13741310 .

Could you please rebase your PR and remove conflicts? I hope it won't be a big issue, but some changes will be needed for sure.  Please let me know once it is done. I will re-run the DB matrix again with your changes - hope it will pass next time.

Thanks again and sorry for the additional work.

**Comment by rmartinc** [comment]
Hi @mposolda 
It's done. I have reduced the number of threads in the test with two level because I've seen that sometimes it gives errors because of timeout requesting new connections to the database (only 20 connections by default).
Regarding your previous changes, I have done the same you did for galera in my `currentIdsInDatabaseChangeLogLockTable` but cannot test.

**Comment by rmartinc** [comment]
Argh! I see that crossdc is not working now. I'll try to check it later today or tomorrow. Sorry!

**Comment by mposolda** [comment]
@rmartinc np, Thanks!

**Comment by rmartinc** [comment]
@mposolda I checked yesterday that the crossdc tests failed because of different versions of the h2 library. Once I updated the pom.xml in tests to the same version, the same code started to work (it made no sense, it's the same exact code that is doing the DBLockTest, doing the same). Launch the automatic tests again if you can.

**Comment by mposolda** [comment]
@rmartinc Sorry to bother again, but could you please rebase with latest master? That will allow me to finally run DB tests and hopefully without failure this time. Thanks for the patience.

**Comment by rmartinc** [comment]
Rebased now. It was a clean rebase. Let's see the tests...

**Comment by mposolda** [comment]
@rmartinc Thanks! Running DB matrix now. Will let you know how it goes...

**Review by mposolda — APPROVED** [review]
@rmartinc DB matrix passed with all DBs and travis as well. Thanks again for the patience!

**Pull request merged**
